src/buffer/out/Row.cpp

```diff
@@ -3,29 +3,129 @@
+{
+    for (; first != last; ++first)
+    {
+        *first = val;
+    }
+    return first;
+}
+
+// Same as std::fill_n, but purpose-built for very small `count`
+// where a trivial loop outperforms vectorization.
+template<typename OutIt, typename Diff, typename T>
+constexpr OutIt fill_n_small(OutIt dest, Diff count, const T val)
+{
+    for (; count; --count, ++dest)
+    {
+        *dest = val;
+    }
+    return dest;
+}
+
+// Same as std::copy_n, but purpose-built for very short `count`
+// where a trivial loop outperforms vectorization.
+template<typename InIt, typename Diff, typename OutIt>
+constexpr OutIt copy_n_small(InIt first, Diff count, OutIt dest)
+{
+    for (; count; --count, ++dest, ++first)
+    {
+        *dest = *first;
+    }
+    return dest;
+}
 
 // Routine Description:
 // - constructor
 // Arguments:
-// - rowId - the row index in the text buffer
 // - rowWidth - the width of the row, cell elements
 // - fillAttribute - the default text attribute
-// - pParent - the text buffer that this row belongs to
 // Return Value:
 // - constructed object
-ROW::ROW(const til::CoordType rowId, const til::CoordType rowWidth, const TextAttribute fillAttribute, TextBuffer* const pParent) :
-    _id{ rowId },
-    _rowWidth{ rowWidth },
-    _charRow{ rowWidth, this },
-    _attrRow{ rowWidth, fillAttribute },
-    _lineRendition{ LineRendition::SingleWidth },
-    _wrapForced{ false },
-    _doubleBytePadded{ false },
-    _pParent{ pParent }
+ROW::ROW(wchar_t* charsBuffer, uint16_t* charOffsetsBuffer, uint16_t rowWidth, const TextAttribute& fillAttribute) :
+    _charsBuffer{ charsBuffer },
+    _chars{ charsBuffer, rowWidth },
+    _charOffsets{ charOffsetsBuffer, ::base::strict_cast<size_t>(rowWidth) + 1u },
+    _attr{ rowWidth, fillAttribute },
+    _columnCount{ rowWidth }
 {
+    if (_chars.data())
+    {
+        _init();
+    }
+}
+
+void swap(ROW& lhs, ROW& rhs) noexcept
+{
+    std::swap(lhs._charsBuffer, rhs._charsBuffer);
+    std::swap(lhs._charsHeap, rhs._charsHeap);
+    std::swap(lhs._chars, rhs._chars);
+    std::swap(lhs._charOffsets, rhs._charOffsets);
+    std::swap(lhs._attr, rhs._attr);
+    std::swap(lhs._columnCount, rhs._columnCount);
+    std::swap(lhs._lineRendition, rhs._lineRendition);
+    std::swap(lhs._wrapForced, rhs._wrapForced);
+    std::swap(lhs._doubleBytePadded, rhs._doubleBytePadded);
+}
+
+void ROW::SetWrapForced(const bool wrap) noexcept
+{
+    _wrapForced = wrap;
+}
+
+bool ROW::WasWrapForced() const noexcept
+{
+    return _wrapForced;
+}
+
+void ROW::SetDoubleBytePadded(const bool doubleBytePadded) noexcept
+{
+    _doubleBytePadded = doubleBytePadded;
+}
+
+bool ROW::WasDoubleBytePadded() const noexcept
+{
+    return _doubleBytePadded;
+}
+
+LineRendition ROW::GetLineRendition() const noexcept
+{
+    return _lineRendition;
+}
+
+void ROW::SetLineRendition(const LineRendition lineRendition) noexcept
+{
+    _lineRendition = lineRendition;

```

Comment:
nit: instead of `GetX` / `SetX` can we just have it be `X(value)` and `X()`?

Reply:
I personally prefer Get/Set prefixed methods, because they're easier to find with "dumb" code searchers (like GitHub's code search or our internal code search tool).

But in this case I kept it as Get/Set, simply because I wanted to keep the diff minimal. Those methods aren't new after all.